scripts/pi-hole/php/groups.php

```diff
@@ -0,0 +1,831 @@
+<?php
+/* Pi-hole: A black hole for Internet advertisements
+*  (c) 2019 Pi-hole, LLC (https://pi-hole.net)
+*  Network-wide ad blocking via your own hardware.
+*
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+require_once('auth.php');
+
+// Authentication checks
+if(isset($_POST['token']))
+{
+    check_cors();
+    check_csrf($_POST['token']);
+}
+else
+{
+    log_and_die("Not allowed!");
+}
+
+$reload = false;
+
+require_once("func.php");
+require_once("database.php");
+$GRAVITYDB = getGravityDBFilename();
+$db = SQLite3_connect($GRAVITYDB, SQLITE3_OPEN_READWRITE);
+
+function JSON_success($message = null)
+{
+    header('Content-type: application/json');
+    echo json_encode(array("success" => true, "message" => $message));
+}
+
+function JSON_error($message = null)
+{
+    header('Content-type: application/json');
+    $response = array("success" => false, "message" => $message);
+    if(isset($_POST['action']))
+    {
+        array_push($response, array("action" => $_POST['action']));
+    }
+    echo json_encode($response);
+}
+
+if($_POST['action'] == "get_groups")
+{
+    // List all available groups
+    try
+    {
+        $query = $db->query("SELECT * FROM \"group\";");
```

Comment:
You could use single quotes here instead of double quotes (it will stop you having to escape the quotes as well). You should only really use double quotes if your planning on using a variable as it will make the PHP engine check the string to ensure nothing needs to be interpreted

Reply:
I am using single-quotes in many places, I missed this one (and a few others). Thanks for catching this.